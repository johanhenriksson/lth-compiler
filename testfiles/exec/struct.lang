struct st_type {
    int field;
}

int main() {
    st_type instance; 
    instance.field = 5;
    assert(instance.field == 5);
    return 0;
}
