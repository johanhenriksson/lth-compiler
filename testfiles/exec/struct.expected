; Library Functions
declare i8* @__malloc(i32)
declare void @__delete(i8*)
declare void @print(i64)
declare i64 @read()
declare i64 @truncate(double)
declare double @promote(i64)
declare void @assert(i1)
declare void @except(i1)

%st_type = type {
  i64
}

define i64 @main() {
%instance = alloca %st_type*
%_1 = getelementptr %st_type, %st_type* null, i32 1
%_2 = ptrtoint %st_type* %_1 to i32
%_3 = call i8* @__malloc(i32 %_2)
%_4 = bitcast i8* %_3 to %st_type*
store %st_type* %_4, %st_type** %instance
; IdUse instance
%_5 = load %st_type*, %st_type** %instance
; StructIdUse field
%_6 = getelementptr %st_type, %st_type* %_5, i32 0, i32 0
store i64 5, i64* %_6
; IdUse instance
%_8 = load %st_type*, %st_type** %instance
; StructIdUse field
%_9 = getelementptr %st_type, %st_type* %_8, i32 0, i32 0
%_10 = load i64, i64* %_9
%_7 = icmp eq i64 %_10, 5
call void @assert(i1 %_7)
ret i64 0
}

